wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessRouterServiceImpl.java

```diff
@@ -51,26 +56,64 @@ public class LimitlessRouterServiceImpl implements LimitlessRouterService {
           CACHE_CLEANUP_NANO
       );
 
-  public LimitlessRouterServiceImpl() {
-    this(LimitlessRouterMonitor::new);
+  private static final SlidingExpirationCacheWithCleanupThread<String, List<HostSpec>>
+      limitlessRouterCache =
+      new SlidingExpirationCacheWithCleanupThread<>(
+          x  -> true,
+          x -> {},
+        CACHE_CLEANUP_NANO
+      );
+
+  public LimitlessRouterServiceImpl(final @NonNull PluginService pluginService) {
+    this(pluginService, LimitlessRouterMonitor::new);
   }
 
-  public LimitlessRouterServiceImpl(final LimitlessRouterMonitorInitializer limitlessRouterMonitorInitializer) {
+  public LimitlessRouterServiceImpl(
+      final @NonNull PluginService pluginService,
+      final LimitlessRouterMonitorInitializer limitlessRouterMonitorInitializer) {
+    this.pluginService = pluginService;
     this.limitlessRouterMonitorInitializer = limitlessRouterMonitorInitializer;
+    this.queryHelper = new LimitlessQueryHelper(pluginService);
   }
 
   @Override
   public List<HostSpec> getLimitlessRouters(final String clusterId, final Properties props) throws SQLException {
     final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
         MONITOR_DISPOSAL_TIME_MS.getLong(props));
+    final List<HostSpec> limitlessRouters = limitlessRouterCache.get(clusterId, cacheExpirationNano);
+    if (limitlessRouters == null) {
+      return Collections.emptyList();
+    }
+    return limitlessRouters;
+  }
 
-    final LimitlessRouterMonitor
-        limitlessRouterMonitor = limitlessRouterMonitors.get(clusterId, cacheExpirationNano);
-    if (limitlessRouterMonitor == null) {
-      throw new SQLException(
-          Messages.get("LimitlessRouterServiceImpl.nulLimitlessRouterMonitor", new Object[]{clusterId}));
+  @Override
+  public List<HostSpec> forceGetLimitlessRoutersWithConn(
+      final Connection connection, final int hostPort, final Properties props) throws SQLException {
+    final long cacheExpirationNano = TimeUnit.MILLISECONDS.toNanos(
+        MONITOR_DISPOSAL_TIME_MS.getLong(props));
+
+    forceGetLimitlessRoutersLock.lock();
+    try {
+      final List<HostSpec> limitlessRouters =
+          limitlessRouterCache.get(this.pluginService.getHostListProvider().getClusterId(), cacheExpirationNano);
+      if (!Utils.isNullOrEmpty(limitlessRouters)) {
+        return limitlessRouters;
+      }
+
+      final List<HostSpec> newLimitLessRouters =
+          this.queryHelper.queryForLimitlessRouters(connection, hostPort);
+
+      limitlessRouterCache.remove(this.pluginService.getHostListProvider().getClusterId());
```

Comment:
Why do we need to remove the entry?

Reply:
updated SlidingExpirationCache with `put()` method